### PR TITLE
GenJetFlavourInfos: fix compiler warning, use prunedGenParticles in associations to correctly match pat jet flavour info

### PR DESCRIPTION
**Pull request by Andrej-CMS**
please see #19048 
addressed comment by @davidlange6

### CODE REVIEW

**Comment by perrotta** [comment]
@Andrej-CMS : maybe you can profit and apply also the two small fixes suggested by the code-checker

**Comment by perrotta** [comment]
please test

File: PhysicsTools/PatAlgos/plugins/GenJetFlavourInfoPreserver.cc [hunk 1]
```diff
@@ -72,12 +72,13 @@ pat::GenJetFlavourInfoPreserver::produce(edm::Event & iEvent, const edm::EventSe
 
     edm::Ref<std::vector<reco::GenJet> > slimmedGenJetRef;
 
-    int i = -1;
-    for (auto const& genJet : *genJets){
-         i++;
+
+    for (uint i=0; i<genJets->size();++i){
```
Comment (davidlange6) [inline PhysicsTools/PatAlgos/plugins/GenJetFlavourInfoPreserver.cc]: hi @Andrej-CMS - please change uint to unsigned int

**Comment by Andrej-CMS** [comment]
Hello all,
Checking the code again and producing new MiniAODSim, I realized that I made a mistake. 
When we decided in PR #19048 the genParticles should be used for PF jets, I forgot to change the HadronAndPartonSelector for slimmedGenJetsFlavourInfos. These jets need to use prunedGenParticles, because it is essential to trace back what is the origin of the hadronFlavour and, therefore, we need the RefVectors to the hadrons. The merged version doesn't have the RefVectors.

The last commit fixes the issue. I tested it on 10k QCD events to make sure that the RefVectors exist and work as intended. This brings the PR in the state that was discussed in #19048 . The changes here do not affect the PF jets.
I apologize for the honest mistake! This will probably need some discussion with @slava77 .

Kind regards,
Andrej

File: PhysicsTools/JetMCAlgos/python/AK4GenJetFlavourInfos_cfi.py [hunk 2]
```diff
@@ -1,5 +1,11 @@
 import FWCore.ParameterSet.Config as cms
 from PhysicsTools.JetMCAlgos.AK4PFJetsMCFlavourInfos_cfi import ak4JetFlavourInfos
 
-ak4GenJetFlavourInfos = ak4JetFlavourInfos.clone(jets = "ak4GenJetsNoNu")
+ak4GenJetFlavourInfos = ak4JetFlavourInfos.clone(
+											jets = "ak4GenJetsNoNu",
+											bHadrons = cms.InputTag("selectedHadronsAndPartonsForGenJetsFlavourInfos","bHadrons"),
+    										cHadrons = cms.InputTag("selectedHadronsAndPartonsForGenJetsFlavourInfos","cHadrons"),
+  											partons  = cms.InputTag("selectedHadronsAndPartonsForGenJetsFlavourInfos","physicsPartons"),
+    										leptons  = cms.InputTag("selectedHadronsAndPartonsForGenJetsFlavourInfos","leptons"),
```
Comment (slava77) [inline PhysicsTools/JetMCAlgos/python/AK4GenJetFlavourInfos_cfi.py]: please fix the indentation.
It looks like the problem is with the way how tabs are displayed: for this just replace tabs with spaces

**Comment by slava77** [comment]
@ahinzmann @ferencek 
please check and confirm that the latest update to ak4GenJetFlavourInfos is as desired.
If there may still be some discussion or studies and the version of the code in #19048 is problematic for some use cases, perhaps we should revert the #19048 first

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ahinzmann** [comment]
@Andrej-CMS I am not sure I understand correctly that we have now.
1. Parton and hadron flavors based on genParticles stored in matched GenJets stored in the slimmedJets (with broken references)
2. Parton and hadron flavors based on prunedGenParticles stored as a separate collection ak4GenJetFlavourInfos (with references to hadrons and partons)
Do we have two different flavor informations stored in MiniAOD then?

**Comment by Andrej-CMS** [comment]
@ahinzmann 
There are indeed two different collections saved in the MiniAOD.

1.  the flavour information of slimmedJets based on genParticles with broken references as was desired in #19048 by QGL . There was no use case for the RefVectors of these jets brought up. 

2. the slimmedGenJetsFlavourInfos collection refers to the slimmedGenJets. This collection stores the flavour informations of the ak4GenJets in AOD and maps them to the slimmedGenJets. This information is required for any analysis that relies on knowing the origin of the heavy flavour genJets, e.g. for classifying events on truth level in categories: ttbar+b jet, ttbar+bbbar, etc. 
I have shown in #19048 that reclustering the slimmedGenJets on MiniAOD to get the references to the constituents leads to slightly different results compared to AOD (please see also slide 3 and 4 in my presentation in [1]). This difference leads to the fact that sometimes the heavy flavour gets assigned to a different slimmedGenJet than it would be on AOD or the heavy flavour doesn't get assigned at all. This leads also to a difference between MiniAOD and AOD of about 1% in event yields when requiring two slimmedGenJets that do not come from top quark decay in ttbar+jets samples. To make the slimmedGenJet flavour clustering reproducible on MiniAOD, the slimmedGensJetFlavourInfos branch was added.

If you have any further questions, please let me know!

Kind regards,
Andrej  

[1] https://indico.cern.ch/event/649196/contributions/2641282/attachments/1483715/2302215/JetFlavour.pdf

**Comment by ahinzmann** [comment]
ok!
So I understand standard b-tagging using slimmedJets doesn't need the references, since only the number of b-hadrons in a jet is relevant to apply b-tag scale factors, while truth level splitting of MC samples needs to track in addition from which parton in the hard process the b-hadron actually came from.

Since the q/g part is untouched by this PR, I don't have a strong preference.

@ferencek better comments on the heavy flavor treatment.

**Comment by perrotta** [comment]
please test

**Comment by ferencek** [comment]
There are no issue with this PR from the heavy flavor point of view. This is because b tagging primarily relies on the hadron-based flavor which neither this PR nor #19048 have touched.

The only caveat is that the b-tagging DQM still relies on the parton-based flavor to be able to distinguish different types of light-flavor (non-b and non-c) jets. However, this is done with the option `hadronFlavourHasPriority` enabled in `DQMOffline/RecoB/python/dqmAnalyzer_cff.py` which is different from the default PAT configuration where this option is not enabled. With #19048 merged, the b-tagging DQM further diverges from the default PAT configuration where the parton selection has changed so this is something to keep in mind. For regular users interested in, for example, applying the b-tagging scale factors, what is relevant is the hadron-based flavor so they don't care about these differences.

@imarches, @kskovpen, do you want to add anything?

**Comment by slava77** [comment]
@cmsbuild please test
hopefully  failures in harvesting are transient
@dmitrijus please take a note.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20281	 be715ba
- several fixes
    - technical fix for the compiler warning in GenJetFlavourInfoPreserver.cc; 
    - functional fix to select prunedGenParticles as a source of the jet flavor ID, to be in agreement with the pat jets choice
- jenkins tests pass and comparisons with baseline show differences only in the recently added slimmedGenJetsFlavourInfos parton flavor distributions

@Andrej-CMS please update the PR title to something more descriptive:
e.g. "GenJetFlavourInfos: fix compiler warning, use prunedGenParticles in associations to correctly match pat jet flavour info"

**Comment by davidlange6** [comment]
merge

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 3]
```diff
@@ -326,6 +326,7 @@ def miniAOD_customizeMC(process):
     #GenJetFlavourInfos
     process.load("PhysicsTools.JetMCAlgos.HadronAndPartonSelector_cfi")
     task.add(process.selectedHadronsAndPartons)
```
Comment (ferencek) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: Apologies for not noticing this earlier but with the updates introduced in this PR, is anything consuming `selectedHadronsAndPartons` now? If not, this line can be removed.

**Comment by ferencek** [comment]
@Andrej-CMS, you are claiming the parton flavor for slimmedJets is based on `genParticles` but I still see `genParticles` in https://github.com/Andrej-CMS/cmssw/blob/be715ba8d26c8509e33de8912bff66aa2b3154b9/PhysicsTools/PatAlgos/python/mcMatchLayer0/jetFlavourId_cff.py#L21 overwritten to `prunedGenParticles` in https://github.com/Andrej-CMS/cmssw/blob/be715ba8d26c8509e33de8912bff66aa2b3154b9/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L351 Am I missing something?

**Comment by arizzi** [comment]
@Andrej-CMS  any answer to the comment from @ferencek 

This should be fixed before 940 is out

**Pull request merged**